### PR TITLE
Chapter 4 - completed

### PR DESCRIPTION
**Pull request by lprakashv**
### Solutions for Chapter Four

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +488,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _        = Trap e
+    (<*>) (Reward f) secret = fmap f secret
```
Comment (vrom911) [inline src/Chapter4.hs:495]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -628,8 +664,11 @@ Can you implement a monad version of AND, polymorphic over any monad?
 
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
+
+
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM ma mb = ma >>= (\a -> fmap ((&&) a) mb)
+-- andM ma mb = (fmap (\a -> (&&) a) ma) <*> mb
```
Comment (vrom911) [inline src/Chapter4.hs:671]: The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face: 

If you don't use the `>>=` operator, then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM` due to the nature of the `Applicative`. `Monad` though, can stop when needed, so you can use that for laziness 🙂

File: src/Chapter4.hs [hunk 3]
```diff
@@ -673,6 +712,28 @@ Specifically,
...
+    = EmptyTree
+    | Node a (Tree a) (Tree a)
+
+instance Functor Tree where
+    fmap :: (a -> b) -> Tree a -> Tree b
+    fmap _ EmptyTree    = EmptyTree
+    fmap f (Node v l r) = Node (f v) (fmap f l) (fmap f r)
+
+{-
+Assuming reversing a tree means mirroring it along a vertical axis
```
Comment (vrom911) [inline src/Chapter4.hs:725]: Yes, this is absolutely correct 👍🏼

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work! You finished all 4 Chapters 🥳 
Thank you so much for taking this course! I hope you had fun 🎈 👯  

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

File: src/Chapter4.hs (hunk 2 above)
Comment (lprakashv) [reply src/Chapter4.hs:671]: Okay, a little hard to understand. But, I will try to make the lazier version of this and I guess the `pure` would be helpful with this.

Comment (lprakashv) [reply src/Chapter4.hs]: Done

**Comment by vrom911** [comment]
The solutions look good :+1: I'm going to close this PR, because we can't merge the solutions to our repo 🙂 
Well done, and thank you very much 🤗 
Don't hesitate to reach out if you have any questions!

**Pull request closed without merging**